File: DataStructure-Practice/Session14/Java/WordLadder.java
https://leetcode.com/problems/word-ladder/

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        
        HashSet<String> wordSet = new HashSet<>(wordList);
        if(!wordSet.contains(endWord)) {
            return 0;
        }
        
        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        int ladder = 0; 
        while(queue.size()>0) {
            ladder++; 
            int levelSize = queue.size();
            while(levelSize>0) {
                char[] word = queue.poll().toCharArray();
                for(int i=0; i<word.length; i++) {
                    char originalChar = word[i];
                    for(char c='a'; c<='z'; c++) {
                        word[i] = c;
                        String newWord = new String(word); 
                        if(newWord.equals(endWord)) {
                            return ladder+1;
                        }
                        if(wordSet.contains(newWord)) {    
                            queue.add(newWord);
                            wordSet.remove(newWord);
                        }
                    }
                    word[i] = originalChar;
                }
                levelSize--;
            }
        }
        return 0;
    }
}

File: DataStructure-Practice/Session14/C++/WordLadder.cpp
https://leetcode.com/problems/word-ladder/

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        if(!wordSet.count(endWord)) {
            return 0;
        }
        
        queue<string> q;
        q.push(beginWord);
        int ladder = 0;
        
        while(!q.empty()) {
            ladder++;
            int levelSize = q.size();
            while(levelSize--) {
                string word = q.front(); 
                q.pop();
                for(int i=0; i<word.length(); i++) {
                    char originalChar = word[i];
                    for(char c='a'; c<='z'; c++) {
                        word[i] = c;
                        if(word==endWord) {
                            return ladder+1;
                        }
                        if(wordSet.count(word)>0) {
                            q.push(word);
                            wordSet.erase(word);
                        }
                    }
                    word[i] = originalChar;
                }  
            } 
        }
        return 0;
    }
};